### ISSUE TITLE
[API Proposal]: Add RemoveAt to IList by Index struct overload

### ISSUE THREAD

**Issue by sps014** [dotnet/runtime#58426]
### Background and motivation

Currently, when doing the deletion in IList implemented data structures we do 
```cs
list.RemoveAt(list.Count-1);
//removes last element
```
this syntax is more verbose,  the same can be done as
```cs
list.RemoveAt(^1);
```
Arrays Index already support Index struct
as
```cs
int[] someArray = new int[5] { 1, 2, 3, 4, 5 };
int lastElement = someArray[^1]; // lastElement = 5
```
which is cleaner, so I suggest to please add RemoveAt by index struct overload.

### API Proposal

```C#
namespace System.Collections
{
    public interface IList : ICollection
    {
      
        // Removes the item at position index.
        void RemoveAt(int index);

        //Remove the item at a given position index
        void RemoveAt(Index index);
    }
```     

### API Usage

```C#
list.RemoveAt(^1);
```     

### Risks

This change will require implementing RemoveAt in the following files
```
Array.CoreCLR.cs
Array.cs
ArrayList.cs
List.cs
Collection.cs
ReadOnlyCollection.cs
```

This risk can be avoided if instead of providing an overload in IList, supported Data Structure them self provide overload.

**Comment by GrabYourPitchforks** [comment]
This can also be done as an extension method, no need to change the `IList` interface or add new members to the concrete types.

```cs
public static class CollectionExtensions
{
    public static void RemoveAt(this IList list, Index index) => list.RemoveAt(index.GetOffset(list.Count));
}
```

**Comment by sps014** [comment]
> This can also be done as an extension method, no need to change the `IList` interface or add new members to the concrete types.
> 
> ```cs
> public static class CollectionExtensions
> {
>     public static void RemoveAt(this IList list, Index index) => list.RemoveAt(index.GetOffset(list.Count));
> }
> ```

I know that this can done with extension method, but this will make Index struct more adaptable, for every projects hopefully everyone have not to import such common extension pkgs or classes for such  functionality.

**Comment by huoyaoyuan** [comment]
Adding a member to interface is breaking change. Adding extension method is not.

**Comment by eiriktsarpalis** [comment]
+1 on this needing to be an extension method in `CollectionExtensions`. There would need to be an investigation of other potential methods in `ICollection`/`ICollection<T>`/`IReadOnlyCollection<T>` that might benefit from having an `Index` extension method. @sps014 would you be interested in driving this?

**Comment by sps014** [comment]
Thanks , i am ready for this opportunity .
Will dig deep into it.

**Comment by sps014** [comment]
Some Extension Methods that can be implemented are
| Possible Extension Methods                	| Applicable To                           	|
|------------------------------------	|-----------------------------------------	|
| Insert(Index, T)                   	| Mutable Collections                      	|
| InsertRange(Index, IEnumerable<T>) 	| Mutable Collections                      	|
| IndexOf(T,Index)                   	| Also Applicable to ReadOnly Collections 	|
| RemoveAt(Index)                    	| Mutable Collections                     	|

Here are some Collections that can take advantage of Index Struct

| Collections 	| Extension Method Required 	| Reason (if any) 	|
|-------------	|--------------------	|-----------------	|
| Dictionary  	| No                 	| unsuitable      	|
| HashSet     	| No                 	| unsuitable      	|
| List        	| Yes                	|                 	|
| ArrayList   	| Yes                	|                 	|
| SortedList  	| Yes                	|                 	|
| Stack       	| No                 	| unsuitable      	|
| Queue       	| No                 	| unsuitable      	|
| LinkedList  	| No                 	| unsuitable      	|

Most of the extension methods need not be implemented for Concurrent Collections and IReadOnlyCollections


Kindly feel free to list down other essential methods and collections that have been missed.

**Comment by sps014** [comment]
I think the following functions need an Index based Extension Method,final API should look like 
```cs
public static  class CollectionExtensions
{
        public static void RemoveAt<T>(this List<T> list, Index index);
        public static void Insert<T>(this List<T> list, Index index, T data);
        public static int IndexOf<T>(this List<T> list, T data, Index index);
        public static void InsertRange<T>(this List<T> list, Index index, IEnumerable<T> data);
        public static void RemoveRange<T>(this List<T> list, Index index, int count);
}
```
@eiriktsarpalis sorry for the ping do you have any advice or suggestion?

**Comment by eiriktsarpalis** [comment]
Thanks @sps014, here's a bit of feedback:

- CollectionsExtensions methods generally target interfaces, however your proposed APIs target `List<T>`. Shouldn't these be `IList<T>`?
- Is it possible to derive all of the above methods from `IList<T>` APIs?
- The proposed `RemoveRange` overload could likely be changed to accept a `System.Range` parameter.

**Comment by sps014** [comment]
Thank you so much for your guidance, I will keep the above points in mind.
Currently, it is not possible to derive all methods from `IList<T>` but most of them can be derived.
For example, ` RemoveRange ` will require additional work in the extension method if we derive from `IList`.

Also, I will keep looking for more collection methods where Range can be an argument resulting in less verbose code.

**Comment by sps014** [comment]
These are the Collections Extensions
```cs
public static  class CollectionExtensions
{
        public static void RemoveAt<T>(this IList<T> list, Index index);
        public static void Insert<T>(this IList<T> list, Index index, T data);
        public static int  IndexOf<T>(this IList<T> list, T data, Index index);

        public static void InsertRange<T>(this IList<T> list, Index index, IEnumerable<T> data);
        public static void RemoveRange<T>(this IList<T> list, Range range);
}
```
Here are some possible extensions Methods in Linq
```cs
public static class LinqExtensions
{
      public static TSource? ElementAtOrDefault<TSource> (this IEnumerable<TSource> source, Index index);
      public static IEnumerable<int> Range (Range range);
}
```
I have a few questions:
 since `InsertRange` and `RemoveRange` Methods are entirely absent in `IList<T>` API but they are present in `List<T>`
should I implement `InsertRange<T>(this List<T> list, Index index, IEnumerable<T> data);` and `RemoveRange` for List only or create 
an implementation for `IList<T>`, will that be a breaking change?
Also, should I create another issue for Linq Methods mentioned above?

@eiriktsarpalis @GrabYourPitchforks any suggestions

**Comment by eiriktsarpalis** [comment]
> since InsertRange and RemoveRange Methods are entirely absent in IList<T> API but they are present in List<T>
should I implement InsertRange<T>(this List<T> list, Index index, IEnumerable<T> data); and RemoveRange for List only or create

I think this would suggest we shouldn't be adding these particular methods to `CollectionExtensions` at all.

**Comment by sps014** [comment]
Leaving LinqExtensions , `InsertRange` and `RemoveRange` can rest of 3 extension methods API be approved for implementation ?

**Comment by sps014** [comment]
sorry for ping @eiriktsarpalis , can these 3 methods be approved for implementation?
these 3 do not introduce any breaking changes, as `int` variant of these methods are already available in `IList<T>`  API.
```cs
public static  class CollectionExtensions
{
        public static void RemoveAt<T>(this IList<T> list, Index index);
        public static void Insert<T>(this IList<T> list, Index index, T data);
        public static int  IndexOf<T>(this IList<T> list, T data, Index index);
}
```

**Comment by eiriktsarpalis** [comment]
What would the `Index` parameter do in the `IndexOf` method? 

For an API proposal to be successful, we would need to also see a few code examples that demonstrate what the new APIs are useful for. Please also see out [API review process](https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md) on more details on what would be required.

**Comment by sps014** [comment]
Suppose we have a list as declared below, with help of these extensions we will be able to do the following
```cs

var list=new List<int>{1,4,3,4,7,1,4,6};

list.Insert(list.Count-1, 5);   //1 4 3 4 7 1 4 5 6
// or  
list.Insert(^1, 5);             //1 4 3 4 7 1 4 5 6


//to remove 2nd last for example we had to do
list.RemoveAt(list.Count-2);   //1 4 3 4 7 1 4 6
//new method reduces it to
list.RemoveAt(^2);             //1 4 3 4 7 1 4 6



//to get the next index from two positions from end we had to write previously
list.IndexOf(4,list.Count-2);  //6
//new method reduces it to 
list.IndexOf(4,^2);            //6


```

**Comment by eiriktsarpalis** [comment]
Sure, but that would require having an `IList<T>.IndexOf(T item, int index)` overload. I don't believe `IList<T>` has this.

**Comment by sps014** [comment]
Yes `IList<T>` doesn't have this but the `List<T>` has so I thought to utilize it there and implement it for `List<T>`.

**Comment by sps014** [comment]
@eiriktsarpalis do have any suggestion?

**Comment by eiriktsarpalis** [comment]
Per my previous recommendation, I don't think targeting `List<T>` would be valid here.

**Comment by sps014** [comment]
Thanks for the suggestion, from the above methods, list only these 2 look suitable for implementation.
```cs
public static  class CollectionExtensions
{
        public static void RemoveAt<T>(this IList<T> list, Index index);
        public static void Insert<T>(this IList<T> list, Index index, T data);
}
```

**Comment by sps014** [comment]
Should we proceed with them?

**Comment by eiriktsarpalis** [comment]
> Should we proceed with them?

cc @layomia @krwq for second opinions. Both methods seem like reasonable extensions for `IList<T>`.

**Comment by sps014** [comment]
thanks @krwq  i have updated the first post.

**Comment by krwq** [comment]
@sps014 will you be interested also in implementing the feature assuming it gets approved?

**Comment by sps014** [comment]
Yes, i am ready to implement once API get approved.

**Comment by sps014** [comment]
@krwq  Will be glad if it get assigned to me.

**Comment by eiriktsarpalis** [comment]
@krwq will be championing the API proposal during API review. We will assign it back to you assuming it gets approved.

**Comment by krwq** [comment]
We should also consider adding doing this for ICollection and also some instance members i.e. List<T>.InsertRange and possibly other similar on other collections.

**Comment by sps014** [comment]
I was thinking to create separate Issues for them , there are many methods that in Collections that can take advantage of Index and Range.

**Comment by terrajobst** [comment]
* The concept makes sense, but we should avoid forcing consumers through interface dispatch
    - At least for low-overhead types such as `List<T>` we should consider adding matching instance methods
    - Also, are these really all the ones that are missing?
    - Alternatively, could this be done by the compiler?

```C#

namespace System.Collections.Generic
{
    public static  class CollectionExtensions
    {
        public static void RemoveAt<T>(this IList<T> list, Index index);
        public static void Insert<T>(this IList<T> list, Index index, T data);
    }
}
```

**Comment by krwq** [comment]
We've just finished the API review and conclusion is following:

We need to identify all APIs which might benefit from Index overload (i.e. grep for `int index`, `int idx` and similar in the ref files)

Based on the number of findings we'll decide if we prefer overloads or we will ask compiler to add a feature which will automatically do the work for us (i.e. adding attribute to an index parameter, like i.e. `void InsertAt([Indexer] int index, ...).

**Comment by sps014** [comment]
Yes Attribute makes sense if we need to do it for many methods,making compiler do the job makes our life easier, there are hundreds of method in runtime and also in 3rd party that can take advantage of an attribute. I will explore other possibilities and will update as soon as possible.
Thanks for this suggestion.

**Comment by krwq** [comment]
Thanks for all the help in this subject @sps014!

**Comment by sps014** [comment]
I  did  some benchmarks  where I had deleted 100_000_000 numbers from the list from the last index by using  2 methods
```cs
            list.RemoveAt(list.Count-1); //int version
            list.RemoveAt(^1); //index extension method
  ```
There is some overhead for Index Extension Method.

.NET SDK=6.0.100-rc.2.21420.16

|             Method |       Mean |    Error |   StdDev |
|------------------- |-----------:|---------:|---------:|
|   RemoveIntVersion |   996.8 ms | 19.51 ms | 22.46 ms |
| RemoveIndexVersion | 1,082.2 ms | 13.84 ms | 10.80 ms |


We need to generate code from
```cs

   public class List<T> : IList<T>, System.Collections.IList, IReadOnlyList<T>
    {
        public void RemoveAt([Indexer(Count)] int index)
        {
        }
    }
```

to 
```cs
public class List<T> : IList<T>, System.Collections.IList, IReadOnlyList<T>
    {
        public void RemoveAt([Indexer(Count)] int index)
        {
        }
        public void RemoveAt(Index index)
        {
           RemoveAt(index.GetOffset(Count));
        }
    }
```
 Source Generators are not capable of this task, as they can't append to the existing files, only `partial` classes can benefit from SG.
Is it advisable to Explore MsBuild Task to generate the same?

**Comment by eiriktsarpalis** [comment]
Related to https://github.com/dotnet/runtime/issues/321, although that proposal concerns itself with the concrete `List<T>` type.

**Comment by sps014** [comment]
I can write a source generator that can generate Extension method for a given Function.
Consider a class 
```cs
public  class SomeClass
{
    public void SomeMethod([Indexer] int idx){}
}
```
with Source Generator 
we can generate the following extension method during compileTime
```cs 
public static class SomeClassIndexerExtension
{
        public static void SomeMethod(this SomeClass c,Index idx){
             c.SomeMethod(index.GetOffset(c.Count);
         }
}
```

Should i start with this or Any other suggestion @eiriktsarpalis

**Comment by CyrusNajmabadi** [comment]
Note: i'd be amenable to a potential language feature based on the following concept:

If the type is Countable: https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md#adding-index-and-range-support-to-existing-library-types

Then any method which takes an `int` argument also allows for that method to be called using a `System.Index` argument as well.  Compiler can just then translate something like `x.RemoveAt(idx)` into `x.RemoveAt(y.GetOffset(x.Length))`. 

Effectively, we'd take what we did with special casing indexers, and apply that to any method in Countable types.

**Comment by CyrusNajmabadi** [comment]
Proposal on the language side added here: https://github.com/dotnet/csharplang/issues/5596

**Comment by nathan130200** [comment]
That extensions for Index must be present in all enumerated types.

**Comment by sps014** [comment]
> That extensions for Index must be present in all enumerated types.

As far as i know this is going to be language design feature. @CyrusNajmabadi has created a proposal, i suggest you to mention there as well.
